src/arp.rs

```diff
@@ -1,65 +1,189 @@
+    match set(packet, 20, 2, v as u64) {
+        Ok(v) => Ok(v),
+        Err(_) => Err("ARP: set_op_code() failed"),
+    }
+}
 
-        // target MAC
-        copy_nonoverlapping(packet.buffer.offset((packet.start) as isize), packet.buffer.offset((packet.start + 32) as isize), MAC_ADDRESS_LEN);
+pub fn get_src_hw(packet: &PvPacket) -> Result<u64, &str> {
+    match get(packet, 6, 6) {
+        Ok(v) => Ok(v),
+        Err(_) => Err("ARP: get_src_hw() failed"),
+    }
+}
+
+pub fn set_src_hw(packet: &mut PvPacket, v: u64) -> Result<&PvPacket, &str> {
+    match set(packet, 6, 6, v) {
+        Ok(v) => {}
+        Err(_) => {
+            return Err("ARP: set_src_hw() failed");
+        }
+    }
+    match set(packet, 22, 6, v as u64) {
+        Ok(v) => Ok(v),
+        Err(_) => Err("ARP: set_src_hw() failed"),
+    }
+}
+
+pub fn get_src_proto(packet: &mut PvPacket) -> Result<u32, &str> {
+    match get(packet, 28, 4) {
+        Ok(v) => Ok(v as u32),
+        Err(_) => Err("ARP: get_src_proto() failed"),
+    }
+}
 
-        // dest IP (10.0.0.5)
-        std::ptr::write(packet.buffer.offset((packet.start + 38) as isize), 0x0a);
-        std::ptr::write(packet.buffer.offset((packet.start + 39) as isize), 0x00);
-        std::ptr::write(packet.buffer.offset((packet.start + 40) as isize), 0x00);
-        std::ptr::write(packet.buffer.offset((packet.start + 41) as isize), 0x05);
+pub fn set_src_proto(packet: &mut PvPacket, v: u32) -> Result<&PvPacket, &str> {
+    match set(packet, 28, 4, v as u64) {
+        Ok(v) => Ok(v),
+        Err(_) => Err("ARP: set_src_proto() failed"),
+    }
+}
 
-        // // set packet length of ARP response as 42 bytes
-        packet.end = packet.start + ARP_RSPN_PACKET_SIZE;
+pub fn get_dst_hw(packet: &PvPacket) -> Result<u64, &str> {
+    match get(packet, 0, 6) {
+        Ok(v) => Ok(v),
+        Err(_) => Err("ARP: get_dst_hw() failed"),
+    }
+}
+
+pub fn set_dst_hw(packet: &mut PvPacket, v: u64) -> Result<&PvPacket, &str> {
+    match set(packet, 0, 6, v) {
+        Ok(v) => {}
+        Err(_) => {
+            return Err("ARP: set_dst_hw() failed");
+        }
+    }
+    match set(packet, 32, 6, v) {
+        Ok(v) => Ok(v),
+        Err(_) => Err("ARP: set_dst_hw() failed"),
     }
 }
+
+pub fn get_dst_proto(packet: &PvPacket) -> Result<u32, &str> {
+    match get(packet, 38, 4) {
+        Ok(v) => Ok(v as u32),
+        Err(_) => Err("ARP: get_dst_proto() failed"),
+    }
+}
+
+pub fn set_dst_proto(packet: &mut PvPacket, v: u32) -> Result<&PvPacket, &str> {
+    match set(packet, 38, 4, v as u64) {
+        Ok(v) => Ok(v),
+        Err(_) => Err("ARP: set_dst_proto() failed"),
+    }
+}
+
+pub fn gen_arp_response_packet(packet: &mut PvPacket, src_mac_address: &MacAddr) {
+    const ARP_RSPN_PACKET_SIZE: u32 = 42;
+
+    // copy dest MAC
+    set_dst_hw(packet, get_src_hw(packet).unwrap()).expect("ARP: gen_arp_response_packet() failed");
```

Comment:
Calling `.expect()` is dangerous in that the application could be `panicked` while running.
if it's hard to make error handlings for lots of case, what about using error propagation `?` and making return type of `gen_arp_response_packet()`?

using `.expect()` should be used while developing or debugging